Kudu.Core/Deployment/StructuredTextDocument.cs

```diff
@@ -0,0 +1,119 @@
+
+                var entryDepth = StructuredTextDocument.GetEntryDepth(lines[i]);
+                if (entryDepth == depth)
+                {
+                    logEntry = new StructuredTextDocumentEntry<T>
+                    {
+                        LogEntry = _deserializer(lines[i].Substring(depth)),
+                        Children = Enumerable.Empty<StructuredTextDocumentEntry<T>>()
+                    };
+                    collection.Add(logEntry);
+                }
+                else if (entryDepth > depth)
+                {
+                    if (logEntry == null)
+                    {
+                        throw new FormatException(string.Format("log file '{0}' is in an invalid format", _path));
+                    }
+
+                    logEntry.Children = ParseLines(lines, i, entryDepth);
+                    i += (logEntry.Children.Count() - 1);
+                }
+                else if (entryDepth < depth)
+                {
+                    return collection;
+                }
+            }
+            return collection;
+        }
+
+        private void Log(string logEntry)
+        {
+            lock(DocumentLock)
+            {
+                FileSystemHelpers.AppendAllTextToFile(_path, string.Concat(logEntry, System.Environment.NewLine));
+            }
+        }
+
+        private static void ValidateStringMessage(string message)
+        {
+            if (message.StartsWith(StructuredTextDocument.LeadingCharacter, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new FormatException(
+                    string.Format("Serialized log of type '{0}' => '{1}' can't start with character '{2}'",
+                    typeof(T).Name,
+                    message,
+                    StructuredTextDocument.PrintableLeadingCharacter));
+            }
+
+            foreach (var pair in StructuredTextDocument.NotAllowedSequances)
+            {
+                if (message.IndexOf(pair.Key, StringComparison.OrdinalIgnoreCase) != -1)
+                {
+                    throw new FormatException(
+                        string.Format("Serialized log of type '{0}' => '{1}' contains an invalid sequance '{2}' for a logEntry in a StructuredTextDocument",
+                        typeof(T).Name,
+                        message,
+                        pair.Value));
+                }
+            }
+        }
+    }
+
+    public static class StructuredTextDocument
+    {
+        public const string LeadingCharacter = "\t";
+        public const string PrintableLeadingCharacter = "\\t";
+        public static readonly IDictionary<string, string> NotAllowedSequances = new Dictionary<string, string> {  { "\r\n", "\\r\\n" }, { "\r", "\\r" }, { "\n", "\\n" }};
+        public static int GetEntryDepth(string entry)
+        {
+            return entry.TakeWhile(c => LeadingCharacter.Equals(c.ToString(), StringComparison.OrdinalIgnoreCase)).Count();
```

Comment:
"StringComparison.OrdinalIgnoreCase" seems unnecessary, i don`t think "\t" should equal to "\T"